### ISSUE TITLE
Node out of sync and can't recover

### ISSUE THREAD

**Issue by garyyu** [mimblewimble/grin#3030]
A long standing bug come up to the surface.

If the Body Head happen to be a fork, there's no chance to recover by itself except removing the whole `chain_data` and restarting the node.

```Rust
DEBUG grin_servers::common::hooks - Received block header 000002121a63 at 336847 from 139.162.168.18:3414, going to process.
DEBUG grin_chain::pipe - header_head updated to 000002121a63 at 336847

DEBUG grin_chain::pipe - pipe: process_block 000002121a63 at 336847 [in/out/kern: 3/5/3]
DEBUG grin_chain::pipe - head updated to 000002121a63 at 336847

    # now the Body Head is on the fork block 336847@000002121a63
    # then it will reject all the new blocks because it didn't see the main chain block 336847@0000018e109d, before it start a new Syncing.

DEBUG grin_servers::common::hooks - Received block header 000000393005 at 336848 from 139.162.168.18:3414, going to process.
DEBUG grin_servers::common::adapters - Block header 000000393005 refused by chain: StoreErr(NotFoundErr("BLOCK HEADER: 0000018e109d"), "NotFoundErr(\"BLOCK HEADER: 0000018e109d\")")

DEBUG grin_servers::common::hooks - Received block header 000002a11b76 at 336849 from 139.162.168.18:3414, going to process.
DEBUG grin_servers::common::adapters - Block header 000002a11b76 refused by chain: StoreErr(NotFoundErr("BLOCK HEADER: 000000393005"), "NotFoundErr(\"BLOCK HEADER: 000000393005\")")

DEBUG grin_servers::common::hooks - Received block header 000020b1eb94 at 336850 from 139.162.168.18:3414, going to process.
DEBUG grin_servers::common::adapters - Block header 000020b1eb94 refused by chain: StoreErr(NotFoundErr("BLOCK HEADER: 000002a11b76"), "NotFoundErr(\"BLOCK HEADER: 000002a11b76\")")

        ... ...

DEBUG grin_servers::common::hooks - Received block header 000000606f52 at 336853 from 217.8.61.200:3414, going to process.
DEBUG grin_servers::common::adapters - Block header 000000606f52 refused by chain: StoreErr(NotFoundErr("BLOCK HEADER: 000000d2f919"), "NotFoundErr(\"BLOCK HEADER: 000000d2f919\")")

    # now, it enable the Syncing as designed.

INFO grin_servers::grin::sync::syncer - sync: total_difficulty 1024223429718561, peer_difficulty 1024239491668899, threshold 16032338146 (last 5 blocks), enabling sync
DEBUG grin_servers::grin::sync::header_sync - sync: initial transition to HeaderSync. sync_head: 00000173aabd at 336839, resetting to: 000002121a63 at 336847


DEBUG grin_servers::grin::sync::header_sync - sync: request_headers: asking 35.241.140.101:3414 for headers, [000002121a63, 00000196fbc9, 0000024a5c88, 00001c600fa2, 000013933daa, 00000110dd4a, 000001f394f3, 000001300a5a, 000000a4b423, 00002138c3ac, 00000143a5cb, 0000019c32c2, 000001094ec4, 000000615080, 000012b9d5dc, 000000b18adb, 000039ddef02, 0000005112d3, 40adad0aec27]
INFO grin_servers::common::adapters - Received 8 block headers from 35.241.140.101:3414

DEBUG grin_chain::pipe - sync_head updated to 000000606f52 at 336853
DEBUG grin_chain::pipe - header_head updated to 000000606f52 at 336853

        # Header Sync is OK

DEBUG grin_servers::grin::sync::body_sync - body_sync: no pending block request, asking more 
DEBUG grin_chain::chain - body_sync: body_head - 000002121a63, 336847, header_head - 000000606f52, 336853, sync_head - 000000606f52, 336853
DEBUG grin_servers::grin::sync::body_sync - block_sync: 336847/336853 requesting blocks [000000393005, 000002a11b76, 000020b1eb94, 0000143fce88, 000000d2f919, 000000606f52] from 6 peers

        # Problem happen here! It will never try to request the main chain block 336847@0000018e109d, then there's no chance to recover by itself.

DEBUG grin_chain::chain - process_block: orphan: 000000d2f919, # orphans 1
DEBUG grin_chain::chain - process_block: orphan: 000002a11b76, # orphans 2
DEBUG grin_chain::chain - process_block: orphan: 0000143fce88, # orphans 3
DEBUG grin_chain::chain - process_block: orphan: 000000606f52, # orphans 4
DEBUG grin_chain::chain - process_block: orphan: 000000393005, # orphans 5
DEBUG grin_chain::chain - process_block: orphan: 000020b1eb94, # orphans 6
```

The fix solution should be easy, just on the initial transition to HeaderSync state (from NoSync state), when we `self.chain.reset_sync_head()`, we also do a `reset_body_head()` which force the Body Head going back to the fork point.

I will try a fixing PR on that soon.

**Comment by jyap808** [comment]
Cool, looks to be my reported issue on Keybase grincoin.teams.node_dev#general

**Comment by garyyu** [comment]
(It's a bad idea to have a `reset_body_head()`, actually that would need the chain rewinding instead of a simple reset.)

The root cause could be the `is_on_current_chain()` function, which looks like the behavior is not as the expected, to be discussed with @antiochp 

```Rust
	/// Verifies the given block header is actually on the current chain.
	/// Checks the header_by_height index to verify the header is where we say
	/// it is
	pub fn is_on_current_chain(&self, header: &BlockHeader) -> Result<(), Error> {
		let chain_header = self.get_header_by_height(header.height)?;
		if chain_header.hash() == header.hash() {
			Ok(())
		} else {
			Err(ErrorKind::Other(format!("not on current chain")).into())
		}
	}
```
We expect it `verifies the given block header is actually on the current chain`, but it now only verify the header chain?

And `check_txhashset_needed()` use it to decide which blocks to be requested for body sync:
```Rust
		while let Ok(header) = current {
			// break out of the while loop when we find a header common
			// between the header chain and the current body chain
			if header.height <= body_head.height {
				if let Ok(_) = self.is_on_current_chain(&header) {
					break;
				}
			}
```

**Comment by antiochp** [comment]
Hmm this is weird - "body head" should never be a fork - we only ever update `head` if this block is the most work. So it would need to be the first block seen with a certain total difficulty and this would need to be greater than any previous block on that chain or any other known fork.

**Comment by garyyu** [comment]
> What is 336847@0000018e109d?

336847@0000018e109d is the main chain. check https://grinscan.net/block/336847

>  There is no logs to indicate what this is.

No logs. Because this node never saw it at that moment.

**Comment by antiochp** [comment]
> We expect it verifies the given block header is actually on the current chain, but it now only verify the header chain?

Yes. By design. 
We use the header MMR implicitly as the "header by height" index.

But - we can "rewind and reapply" headers to the header MMR to put it in a variety of states based on what we want to do with it (i.e. let it track a temporary fork etc.)

**Comment by garyyu** [comment]
Closed by #3033
